tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,517 @@
+//! This module provides [`LeaderTpuCacheService`] structure along with
+//! [`LeaderUpdateReceiver`], [`LeaderTpuCacheServiceConfig`].
+//! [`LeaderTpuCacheService`] tracks the current and upcoming Solana leader
+//! nodes and their TPU socket addresses.
+#![allow(clippy::arithmetic_side_effects)]
+use {
+    crate::{
+        connection_workers_scheduler::extract_send_leaders,
+        logging::{debug, error, info, warn},
+        node_address_service::{slot_receiver::EstimatedSlot, SlotReceiver},
+    },
+    async_trait::async_trait,
+    solana_clock::{Slot, NUM_CONSECUTIVE_LEADER_SLOTS},
+    solana_pubkey::Pubkey,
+    solana_quic_definitions::QUIC_PORT_OFFSET,
+    solana_rpc_client::nonblocking::rpc_client::RpcClient,
+    solana_rpc_client_api::{client_error::Error as ClientError, response::RpcContactInfo},
+    std::{collections::HashMap, future::Future, net::SocketAddr, str::FromStr, sync::Arc},
+    thiserror::Error,
+    tokio::{
+        sync::watch,
+        task::JoinHandle,
+        time::{interval, sleep, Duration, Instant},
+    },
+    tokio_util::sync::CancellationToken,
+};
+
+/// Maximum number of slots used to build TPU socket fanout set
+const MAX_FANOUT_SLOTS: u64 = 100;
+
+/// Configuration for the [`LeaderTpuCacheService`].
+#[derive(Debug, Clone)]
+pub struct LeaderTpuCacheServiceConfig {
+    /// max number of leaders to look ahead for, not necessary unique.
+    pub lookahead_leaders: u64,
+    /// how often to refresh cluster nodes info.
+    pub refresh_nodes_info_every: Duration,
+    /// maximum number of consecutive failures to tolerate.
+    pub max_consecutive_failures: usize,
+}
+
+/// [`LeaderTpuCacheService`] is a background task that tracks the current and
+/// upcoming Solana leader nodes and updates their TPU socket addresses
+/// encapsulated in [`LeaderUpdateReceiver`] for downstream consumers.
+pub struct LeaderTpuCacheService {
+    handle: Option<JoinHandle<Result<(), LeaderTpuCacheServiceError>>>,
+    cancel: CancellationToken,
+}
+
+/// Receiver for leader TPU socket address updates from
+/// [`LeaderTpuCacheService`].
+#[derive(Clone)]
+pub struct LeaderUpdateReceiver {
+    receiver: watch::Receiver<(Vec<SocketAddr>, bool)>,
```

Comment:
*minor*

It would not hurt to add a struct to give this `bool` a name.

Maybe something like

```rust
struct LeaderChange {
    future_leaders: Vec<SocketAddr>,
    /// TODO: I actually do not undestand what this `bool` is for yet.  So a short notice here could
    /// be quite helpful.
	extended: bool,
}
```

```suggestion
    receiver: watch::Receiver<LeaderChange>,
```